Update: change default role for custom elements

scottaohara opened the pull request:
closes #2303

This updates a custom element's default role to none to allow attribute reflection from the custom element parent to its internals.

This change provides additional clarification about how a custom element can be provided a role by authors, and what caveats would change a custom element's default role of none, to another implicit minimum role.


# Test, Documentation and Implementation tracking
Once this PR has been reviewed and has consensus from the working group, tests should be written and issues should be opened on browsers. Add N/A and check when not applicable.

* [ ] "author MUST" tests:
* [ ] "user agent MUST" tests:
* [ ] Browser implementations (link to issue or commit):
   * WebKit:
   * Gecko:
   * Blink:
* [ ] Does this need AT implementations?
* [ ] Related APG Issue/PR:
* [ ] MDN Issue/PR:

scottaohara committed:
Update: change default role for custom elements

closes #2303

This updates a custom element's default role to none to allow attribute reflection from the custom element parent to its internals.

This change provides additional clarification about how a custom element can be provided a role by authors, and what caveats would change a custom element's default role of none, to another implicit minimum role.

scottaohara commented:
marking as draft as the form associated custom element mapping table also needs updating - but i want to make sure i got this going in the right direction before i edit that table too

scottaohara committed:
fix typo / missing in-page link id

scottaohara commented on html-aam/index.html:
```diff
@@ -849,8 +862,30 @@ <h4 id="el-autonomous-custom-element">autonomous custom element</h4>
+              <th id=custom-element-comments>Comments</th>
+              <td>
+                <p>
+                  Along with the attributes defined in this specification that contribute to modifying an element's <a>minimum role</a>,
```
```suggestion
                  Along with the attributes defined in this specification that participate in modifying an element's <a>minimum role</a>,
```

scottaohara committed:
Update html-aam/index.html

aleventhal commented on html-aam/index.html:
```diff
@@ -849,8 +862,30 @@ <h4 id="el-autonomous-custom-element">autonomous custom element</h4>
+                  <li>if an author specifies an ARIA relation via an element attribute: expose <a class="core-mapping" href="#role-map-generic">`generic`</a> role</li>
+                </ul>
+                <div class="example">
+                  <p>The following is an exmaple of an author specifiying an ARIA relation on a custom element, with another element in the DOM.</p>
```
ex-maple, not ex-oak?

aleventhal replied:
Seriously, is this example for the previous bullet? Because aria-details won't change it from 'none', unless you do myCustomElement.ariaDetailsElements = [ el1, ... ];

scottaohara replied:
ah ok.  thanks.  I didn't realize you were saying that this would only happen if it was set via IDL.

my thought was that this was another use case that didn't make sense to reflect, since the ID that was being referenced would likely be in the light DOM, so it wouldn't work if reflected into the custom element's shadow dom parts

aleventhal commented on html-aam/index.html:
```diff
@@ -8680,7 +8715,7 @@ <h4 id="att-autofocus">`autofocus`</h4>
             <tr>
               <th>Comments</th>
-              <td>Provides a <a href="termref">minimum role</a> of <a class="core-mapping" href="#role-map-group">`group`</a>.</td>
+              <td>Provides a <a>minimum role</a> of <a class="core-mapping" href="#role-map-group">`group`</a>.</td>
```
Did you intend to remove href="termref" here and in the other place?

scottaohara replied:
yes.  it was broken, this fixed the link

scottaohara commented on html-aam/index.html:
```diff
@@ -849,8 +862,30 @@ <h4 id="el-autonomous-custom-element">autonomous custom element</h4>
+                </ul>
+                <div class="example">
+                  <p>The following is an exmaple of an author specifiying an ARIA relation on a custom element, with another element in the DOM.</p>
+                  <pre>
+                    <code>
```
```suggestion
                  <pre><code>
```

scottaohara commented on html-aam/index.html:
```diff
@@ -849,8 +862,30 @@ <h4 id="el-autonomous-custom-element">autonomous custom element</h4>
+                      &lt;my-element aria-details=ex>...&lt;my-element>
+                      ...
+                      &lt;div id=ex> &lt;!-- details for my-element --> &lt;/div>
+                    </code>
+                  </pre>
```
```suggestion
                  </code></pre>
```

scottaohara committed:
Update html-aam/index.html

scottaohara committed:
Update html-aam/index.html

scottaohara committed:
update per feedback/discussion about referenced elements

scottaohara committed:
Update index.html

scottaohara commented on html-aam/index.html:
```diff
@@ -849,8 +862,38 @@ <h4 id="el-autonomous-custom-element">autonomous custom element</h4>
+    &lt;id-lement> ... &lt;/id-lement>
+    &lt;script>
+      ...
+      document.querySelector('id-lement').ariaDescribedby = anotherEl;
```
```suggestion
      document.querySelector('id-lement').ariaDescribedbyByElements = [ descElement ];
```

scottaohara commented on html-aam/index.html:
```diff
@@ -849,8 +862,38 @@ <h4 id="el-autonomous-custom-element">autonomous custom element</h4>
+                  <li>if the custom element is focusable: expose as <a class="core-mapping" href="#role-map-group">`group`</a> role</li>
+                  <li>if the custom element has an `aria-live` attribute: expose as <a class="core-mapping" href="#role-map-generic">`generic`</a> role</li>
+                  <li>if element internals are used to set a global ARIA property: expose as <a class="core-mapping" href="#role-map-generic">`generic`</a> role</li>
+                  <li>if an author specifies an ARIA attribute that creates a relation with another element: expose <a class="core-mapping" href="#role-map-generic">`generic`</a> role</li>
```
```suggestion
                  <li>if an author specifies an ARIA attribute on the custom element that creates a relation with another element: expose <a class="core-mapping" href="#role-map-generic">`group`</a> role</li>
```

scottaohara committed:
Update html-aam/index.html

scottaohara committed:
Update html-aam/index.html

scottaohara committed:
or else if otherwise sure?

scottaohara committed:
combining things

pull comments into role table cell
combine custom elements and form associated custom elements into a single mapping table

aleventhal reviewed (approved):
Awesome, after many iterations

scottaohara committed:
add in no shadow root use cae

update wording to be consistent between steps

scottaohara committed:
Update html-aam/index.html

Co-authored-by: James Craig <cookiecrook@users.noreply.github.com>

scottaohara commented on html-aam/index.html:
```diff
@@ -804,18 +805,35 @@ <h4 id="el-audio">`audio`</h4>
+                  <li>else if the custom element is focusable: map to the <a class="core-mapping" href="#role-map-generic">`group`</a> role</li>
+                  <li>else if an author specifies a global ARIA attribute on the custom element that creates a relation with another element: map to the <a class="core-mapping" href="#role-map-generic">`group`</a> role</li>
+                  <li>else if the custom element has no attached shadow root: map to the <a class="core-mapping" href="#role-map-generic">`generic`</a> role</li>
+                  <li>else if the custom element has an `aria-live` attribute: map to the <a class="core-mapping" href="#role-map-generic">`generic`</a> role</li>
```
i'm not sure i follow this comment - since custom elements are presently generic anything that says to map this to generic is essentially keeping the status quo.

scottaohara committed:
Update html-aam/index.html

Co-authored-by: James Craig <cookiecrook@users.noreply.github.com>

scottaohara commented on html-aam/index.html:
```diff
@@ -226,7 +226,8 @@ <h3>HTML Element Role Mappings</h3>
           </li>
           <li>
-            Where an element is indicated as having &quot;No corresponding (WAI-ARIA) role&quot;, or is mapped to the <a class="core-mapping" href="#role-map-generic">`generic`</a> role, user agents
+            Where an element is indicated as having &quot;No corresponding (WAI-ARIA) role&quot;, or is mapped to the <a class="core-mapping" href="#role-map-generic">`generic`</a>
```
this text existed prior to the creation of the html-* platform roles.  i don't see a problem with revising this to just generic and none.  i think for some platform roles one's milage will vary if anything meaningful actually happens if one were to try and change the role description - but i can definitely see why maybe someone might want to use it on a video element, for instance.

scottaohara committed:
Update html-aam/index.html

Co-authored-by: James Craig <cookiecrook@users.noreply.github.com>

scottaohara commented on html-aam/index.html:
```diff
@@ -804,18 +805,35 @@ <h4 id="el-audio">`audio`</h4>
+                <ul>
+                  <li>If the author assigned a conforming role using the `role` attribute, or by the custom element's internals: map to the specified role.</li>
+                  <li>else if the author assigned HTML attributes that result in a <a>minimum role</a>:  map to the minimum role.</li>
+                  <li>else if the custom element is focusable: map to the <a class="core-mapping" href="#role-map-generic">`group`</a> role</li>
```
```suggestion
                  <li>else if the custom element is focusable: map to the <a class="core-mapping" href="#role-map-group">`group`</a> role</li>
```

scottaohara committed:
Update html-aam/index.html

scottaohara commented on html-aam/index.html:
```diff
@@ -804,18 +805,35 @@ <h4 id="el-audio">`audio`</h4>
+                  <li>else if the author assigned HTML attributes that result in a <a>minimum role</a>:  map to the minimum role.</li>
+                  <li>else if the custom element is focusable: map to the <a class="core-mapping" href="#role-map-generic">`group`</a> role</li>
+                  <li>else if an author specifies a global ARIA attribute on the custom element that creates a relation with another element: map to the <a class="core-mapping" href="#role-map-generic">`group`</a> role</li>
+                  <li>else if the custom element has no attached shadow root: map to the <a class="core-mapping" href="#role-map-generic">`generic`</a> role</li>
```
custom elements are currently generic, so even without this spec change, this use case doesn't work now.

scottaohara commented on html-aam/index.html:
```diff
@@ -804,18 +805,35 @@ <h4 id="el-audio">`audio`</h4>
+                <ul>
+                  <li>If the author assigned a conforming ARIA role using the `role` attribute, or by the custom element's internals: map to the specified role.</li>
+                  <li>else if the author assigned HTML attributes that result in a <a>minimum role</a>:  map to the minimum role.</li>
+                  <li>else if the custom element is focusable: map to the <a class="core-mapping" href="#role-map-generic">`group`</a> role</li>
```
fixed the incorrect link.  i never thought to use rref in html aam - wasn't sure that'd work.  i don't see that used in core aam either.

maybe it'd work now with the specs all in one repo?

re: about mapping to group or something else - i'm open to ideas.  cc @aleventhal if you have thoughts?

spectranaut commented:
@cookiecrook can you respond to @scottaohara comments to help move this forward?

jcsteh commented on html-aam/index.html:
```diff
@@ -804,18 +805,35 @@ <h4 id="el-audio">`audio`</h4>
+                <ul>
+                  <li>If the author assigned a conforming role using the `role` attribute, or by the custom element's internals: map to the specified role.</li>
+                  <li>else if the author assigned HTML attributes that result in a <a>minimum role</a>:  map to the minimum role.</li>
+                  <li>else if the custom element is focusable: map to the <a class="core-mapping" href="#role-map-group">`group`</a> role</li>
```
Should this be done even if a custom element's ShadowRoot has delegatesFocus: true? It's a tricky one. On one hand, it's a pretty important principle to expose focusable elements to accessibility. On the other hand, when delegatesFocus is true, the element isn't in the tab order; it is effectively transparent as far as focus is concerned.

I think it's probably safer to still use role="group" even for delegatesFocus: true. For example, browser native date/time inputs expose the container and allow it to be focused, even though it isn't part of the tab order. Nevertheless, I thought I'd clarify because I haven't seen any discussion of delegatesFocus either here or in #2303.

jnurthen replied:
Good question which we should answer. As delegatesFocus doesn't actually make the shadow host itself "focusable" per the HTML spec's definition (sequentially/click focusable, based on tabindex or native semantics), as worded now, a delegating host with no tabindex of its own would fall through the focusable rule entirely and land on the generic/none rules below so not group.

If we want the group outcome for consistency with native composite controls (date/time inputs, etc.), I think we need to say so explicitly rather than relying on "focusable" to cover it — something like:

"...or is the host of a shadow root whose delegatesFocus is true and which contains a focusable area"

It might also be worth linking "focusable" to [the HTML spec's definition](https://html.spec.whatwg.org/multipage/interaction.html#focusable-area) while we're at it

jcsteh reviewed (commented):
Sorry for the ridiculously slow turnaround here. Just a couple of questions, but overall this looks good to me.

> This updates a custom element's default role to none to allow attribute reflection from the custom element parent to its internals.

Just to be extra clear, when you say "allow attribute reflection" here, you mean that the component could copy attributes itself (e.g. aria-label), not that these would be implicitly reflected somehow? Nothing in the proposed spec text suggests implicit reflectoin; I just wanted to double check my understanding.
